### PR TITLE
chore(RHTAPWATCH-562): Create a dummy service to be deployed in RHTAP

### PR DESCRIPTION
**Pull request by Kousalya1998**
- Add a dummy service in Go
- Add the unit test file in Go

### CODE REVIEW

**Review by hmariset — APPROVED** [review]
LGTM

**Comment by Kousalya1998** [comment]
Reference link:
[1] https://www.digitalocean.com/community/tutorials/how-to-write-unit-tests-in-go-using-go-test-and-the-testing-package
[2] https://wawand.co/blog/posts/how-to-mock-an-external-service/ 
[3] https://www.youtube.com/watch?v=XQzTUa9LPU8

**Review by yftacherzog — CHANGES_REQUESTED** [review]
Nicely done. Please move everything to a sub-directory under the exporter dir (per the ACs). We might have more exporters in the future.

**Comment by Kousalya1998** [comment]
> Nicely done. Please move everything to a sub-directory under the exporter dir (per the ACs). We might have more exporters in the future.


@yftacherzog like o11y/exporter/<all files>

**Comment by yftacherzog** [comment]
> > Nicely done. Please move everything to a sub-directory under the exporter dir (per the ACs). We might have more exporters in the future.
> 
> @yftacherzog like o11y/exporter/ < new files>

Tomorrow we might add another exporter that generates some other metrics.
Something like this:
```
─── exporters
    ├── another-exporter
    │   └── main.go
    └── some-exporter
        └── main.go

```

**Comment by Kousalya1998** [comment]
> > > Nicely done. Please move everything to a sub-directory under the exporter dir (per the ACs). We might have more exporters in the future.
> > 
> > 
> > @yftacherzog like o11y/exporter/ < new files>
> 
> Tomorrow we might add another exporter that generates some other metrics. Something like this:
> 
> ```
> ─── exporters
>     ├── another-exporter
>     │   └── main.go
>     └── some-exporter
>         └── main.go
> ```
> 
> I think it should be `exporters`. i.e. plural

Thanks! Will add it this way.

**Comment by yftacherzog** [comment]
@Kousalya1998 The naming of files/dirs/modules should already be according to the name of the exporter we want to have later on (@hmariset @avi-biton @amisstea please suggest names). This way when we introduce the logic, we change only the logic and not the way we trigger it.

File: exporters/dummy_service_exporter/main_test.go
Comment (Omeramsc) [inline exporters/dummy_service_exporter/main_test.go:28]: Test is mostly self-explanatory, I don't think we need all those comments, or at least most of them.

**Review by Omeramsc — APPROVED** [review]
LGTM

File: exporters/dummy_service_exporter/go.mod
Comment (yftacherzog) [inline exporters/dummy_service_exporter/go.mod:1]: I think we should use segment-bridge as reference here: https://github.com/redhat-appstudio/segment-bridge/blob/main/go.mod

i.e. having go.mod at the top of the repo and naming the module after its location (`github.com/redhat-appstudio/o11y.git`)

File: exporters/dummy_service_exporter/main.go
Comment (yftacherzog) [inline exporters/dummy_service_exporter/main.go:1]: I think we should name the package something like `dsexporter`.

File: exporters/dummy_service_exporter/go.mod
Comment (yftacherzog) [reply exporters/dummy_service_exporter/go.mod:1]: Ultimately, I want for us to be able to run `go test ./...` at the top of the repo and have all test cases executed. I think this should allow that.

**Comment by avi-biton** [comment]
> @Kousalya1998 The naming of files/dirs/modules should already be according to the name of the exporter we want to have later on (@hmariset @avi-biton @amisstea please suggest names). This way when we introduce the logic, we change only the logic and not the way we trigger it.

I would suggest `datasource_exporter` or `grafana_datasource_exporter`
Other than that the file structure looks good

**Comment by yftacherzog** [comment]
> > @Kousalya1998 The naming of files/dirs/modules should already be according to the name of the exporter we want to have later on (@hmariset @avi-biton @amisstea please suggest names). This way when we introduce the logic, we change only the logic and not the way we trigger it.
> 
> I would suggest `datasource_exporter` or `grafana_datasource_exporter` Other than that the file structure looks good

I reading through golang documentation, and as far as I understand it, the package name should be all lowercase without underscores or hyphens, so maybe `dsexporter`.

I think the directory name can be the same. 

I left another comment about the module name. I think we should follow what was done for segment-bridge.

**Comment by avi-biton** [comment]
> > > @Kousalya1998 The naming of files/dirs/modules should already be according to the name of the exporter we want to have later on (@hmariset @avi-biton @amisstea please suggest names). This way when we introduce the logic, we change only the logic and not the way we trigger it.
> > 
> > 
> > I would suggest `datasource_exporter` or `grafana_datasource_exporter` Other than that the file structure looks good
> 
> I'm reading through golang documentation, and as far as I understand it, the package name should be all lowercase without underscores or hyphens, so maybe `dsexporter`.
> 
> I think the directory name can be the same.
> 
> I left another comment about the module name. I think we should follow what was done for segment-bridge.
I am fine with that

File: exporters/dummy_service_exporter/main_test.go
Comment (Kousalya1998) [reply exporters/dummy_service_exporter/main_test.go]: @Omeramsc Updated it.

File: exporters/dummy_service_exporter/go.mod
Comment (Kousalya1998) [reply exporters/dummy_service_exporter/go.mod:1]: @yftacherzog Added. Thanks for the input.

**Comment by Kousalya1998** [comment]
> > @Kousalya1998 The naming of files/dirs/modules should already be according to the name of the exporter we want to have later on (@hmariset @avi-biton @amisstea please suggest names). This way when we introduce the logic, we change only the logic and not the way we trigger it.
> 
> I would suggest `datasource_exporter` or `grafana_datasource_exporter` Other than that the file structure looks good

@avi-biton @yftacherzog

> > > @Kousalya1998 The naming of files/dirs/modules should already be according to the name of the exporter we want to have later on (@hmariset @avi-biton @amisstea please suggest names). This way when we introduce the logic, we change only the logic and not the way we trigger it.
> > 
> > 
> > I would suggest `datasource_exporter` or `grafana_datasource_exporter` Other than that the file structure looks good
> 
> I'm reading through golang documentation, and as far as I understand it, the package name should be all lowercase without underscores or hyphens, so maybe `dsexporter`.
> 
> I think the directory name can be the same.
> 
> I left another comment about the module name. I think we should follow what was done for segment-bridge.

@yftacherzog @Omeramsc Despite changing the file names to dsexporter.go and dsexporter_test.go, when I change the package name from main to something else, I get an error that says:
`package command-line-arguments is not a main package`

File: CONTRIBUTING.md
Comment (hmariset) [inline CONTRIBUTING.md:51]: Go test files are located in each sub-directory under `o11y/exporters` directory with `_test.go` file names.

**Comment by hmariset** [comment]
> > > @Kousalya1998 The naming of files/dirs/modules should already be according to the name of the exporter we want to have later on (@hmariset @avi-biton @amisstea please suggest names). This way when we introduce the logic, we change only the logic and not the way we trigger it.
> > 
> > 
> > I would suggest `datasource_exporter` or `grafana_datasource_exporter` Other than that the file structure looks good
> 
> @avi-biton @yftacherzog
> 
> > > > @Kousalya1998 The naming of files/dirs/modules should already be according to the name of the exporter we want to have later on (@hmariset @avi-biton @amisstea please suggest names). This way when we introduce the logic, we change only the logic and not the way we trigger it.
> > > 
> > > 
> > > I would suggest `datasource_exporter` or `grafana_datasource_exporter` Other than that the file structure looks good
> > 
> > 
> > I'm reading through golang documentation, and as far as I understand it, the package name should be all lowercase without underscores or hyphens, so maybe `dsexporter`.
> > I think the directory name can be the same.
> > I left another comment about the module name. I think we should follow what was done for segment-bridge.
> 
> @yftacherzog @Omeramsc Despite changing the file names to dsexporter.go and dsexporter_test.go, when I change the package name from main to something else, I get an error that says: `package command-line-arguments is not a main package`
> 
> So, I left it as main for now.

Did you change the sub-directory name to `dsexporter`?

**Comment by yftacherzog** [comment]
> @yftacherzog @Omeramsc Despite changing the file names to dsexporter.go and dsexporter_test.go, when I change the package name from main to something else, I get an error that says: package command-line-arguments is not a main package

I'm not versed enough with go to tell what are the best practices here.

@amisstea can you help here?
We want to have an `exporters` directory at the root of the o11y repo and have sub-directories for the individual exporters. 
What are the limitations/best-practices around the names of the files, packages, etc.?
For now we just need a single sub-directory for the current exporter with one file for the code and one file for the test. What are the limitations we have, considering that this code is going to run as part of a deployment?

**Comment by avi-biton** [comment]
@Kousalya1998 Please rename the directory `dummy_service_exporter` to `dsexporter`

**Comment by avi-biton** [comment]
> > > @Kousalya1998 The naming of files/dirs/modules should already be according to the name of the exporter we want to have later on (@hmariset @avi-biton @amisstea please suggest names). This way when we introduce the logic, we change only the logic and not the way we trigger it.
> > 
> > 
> > I would suggest `datasource_exporter` or `grafana_datasource_exporter` Other than that the file structure looks good
> 
> @avi-biton @yftacherzog
> 
> > > > @Kousalya1998 The naming of files/dirs/modules should already be according to the name of the exporter we want to have later on (@hmariset @avi-biton @amisstea please suggest names). This way when we introduce the logic, we change only the logic and not the way we trigger it.
> > > 
> > > 
> > > I would suggest `datasource_exporter` or `grafana_datasource_exporter` Other than that the file structure looks good
> > 
> > 
> > I'm reading through golang documentation, and as far as I understand it, the package name should be all lowercase without underscores or hyphens, so maybe `dsexporter`.
> > I think the directory name can be the same.
> > I left another comment about the module name. I think we should follow what was done for segment-bridge.
> 
> @yftacherzog @Omeramsc Despite changing the file names to dsexporter.go and dsexporter_test.go, when I change the package name from main to something else, I get an error that says: `package command-line-arguments is not a main package`
> 
> So, I left it as main for now.

By reading here https://go.dev/doc/code#:~:text=The%20first%20statement%20in%20a,must%20always%20use%20package%20main%20. It says
`The first statement in a Go source file must be package name. Executable commands must always use package main.`

**Comment by Omeramsc** [comment]
> > > > @Kousalya1998 The naming of files/dirs/modules should already be according to the name of the exporter we want to have later on (@hmariset @avi-biton @amisstea please suggest names). This way when we introduce the logic, we change only the logic and not the way we trigger it.
> > > 
> > > 
> > > I would suggest `datasource_exporter` or `grafana_datasource_exporter` Other than that the file structure looks good
> > 
> > 
> > @avi-biton @yftacherzog
> > > > > @Kousalya1998 The naming of files/dirs/modules should already be according to the name of the exporter we want to have later on (@hmariset @avi-biton @amisstea please suggest names). This way when we introduce the logic, we change only the logic and not the way we trigger it.
> > > > 
> > > > 
> > > > I would suggest `datasource_exporter` or `grafana_datasource_exporter` Other than that the file structure looks good
> > > 
> > > 
> > > I'm reading through golang documentation, and as far as I understand it, the package name should be all lowercase without underscores or hyphens, so maybe `dsexporter`.
> > > I think the directory name can be the same.
> > > I left another comment about the module name. I think we should follow what was done for segment-bridge.
> > 
> > 
> > @yftacherzog @Omeramsc Despite changing the file names to dsexporter.go and dsexporter_test.go, when I change the package name from main to something else, I get an error that says: `package command-line-arguments is not a main package`
> > So, I left it as main for now.
> 
> By reading here https://go.dev/doc/code#:~:text=The%20first%20statement%20in%20a,must%20always%20use%20package%20main%20. It says `The first statement in a Go source file must be package name. Executable commands must always use package main.`

Correct. To put it in simple terms, if your Go file is planned to be an executable (= need to run using a main function), it must be a part of `main` package. otherwise, it can be defined as a different package. so it depends on the use case

**Comment by yftacherzog** [comment]
> I'm not fully aware how will this dummy exporter be used, but it doesn't seem to me like it needs a main.

Our best bet, I think, will be to follow the patterns used on official exporters:

https://github.com/prometheus/haproxy_exporter/tree/main
https://github.com/prometheus/mysqld_exporter
https://github.com/prometheus/node_exporter

Which seems to be `main` package and `main` function.

**Comment by amisstea** [comment]
+1 to renaming the directory to `dsexporter` and sticking with the `main` package. As already mentioned, Go expects the `main` function to be in the `main` package and since this exporter is so simple there's no need to introduce additional packages.

**Comment by yftacherzog** [comment]
> @hmariset @avi-biton Updated the directory name to `dsexporter`. Any suggestions for the file names?

dsexporter.go, dsexporter_test.go

File: CONTRIBUTING.md
Comment (yftacherzog) [inline CONTRIBUTING.md:51]:
```suggestion
This repository contains Go unit tests. Test files are located in the `o11y/exporters` directory with `_test.go` file names.
```

File: CONTRIBUTING.md
Comment (yftacherzog) [inline CONTRIBUTING.md:1]: Let's keep the lines throughout the markdown file shorter (~90 chars).

File: CONTRIBUTING.md
Comment (yftacherzog) [inline CONTRIBUTING.md:53]: I'm not sure what `go clean -testcache` does exactly, but caching exists for a reason and I wouldn't like contributors to clean it every time as a habit.
Eventually, it might mean significantly longer execution times. I don't want all contributors to just use unless it's really needed.

Comment (Kousalya1998) [reply CONTRIBUTING.md:53]: Makes sense. I updated it with just `go test ./...`

**Review by hmariset — APPROVED** [review]

**Comment by amisstea** [comment]
It's nice to see documentation about running the unit tests but there's more that could be added. I think we should have a README in `dsexporter/` explaining what it is and why it can be useful.  There should also be a blurb about exporters in the root README.

**Comment by Kousalya1998** [comment]
> It's nice to see documentation about running the unit tests but there's more that could be added. I think we should have a README in `dsexporter/` explaining what it is and why it can be useful. There should also be a blurb about exporters in the root README.

@amisstea Thank you for the suggestion. It makes sense to describe what it is and how it can be helpful in detail. To create a README inside `dsexporter/` and update o11y/exporter's root README, I created a ticket.

https://issues.redhat.com/browse/RHTAPWATCH-592

**Comment by Kousalya1998** [comment]
@yftacherzog I might need Yftach's approval to merge this ticket.

**Review by amisstea — APPROVED** [review]
Thanks for filing the task @Kousalya1998. I see @yftacherzog's comments were resolved. I'll merge this to unblock our other tasks.

**Pull request merged**
